ledger/internal/double_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package internal_test
+
+import (
+	"testing"
+
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/txntest"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/ledger/internal"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/stretchr/testify/require"
+)
+
+// DoubleLedger allows for easy "Double Entry bookkeeping" as a way to write
+// fairly extensive ledger tests. In addition to simplifiying the addition of
+// txns and txgroups to a ledger (and then allowing for inspection of the
+// created blocks), it also does a double check on correctness by marshalling
+// the created blocks, evaluating the transactions in a ledger copy, and
+// asserting that it comes out the same.  During the insertion of those
+// transactions, the validator ledger is not in `generate` mode - so it
+// evaluates and validates, checking that the ApplyDatas that come from the
+// first ledger match the ADs created by the second. The validator ledger is
+// then temporarily placed in `generate` mode so that the entire block can be
+// generated in the copy second ledger, and compared.
+type DoubleLedger struct {
+	t *testing.T
+
+	generator *ledger.Ledger
+	validator *ledger.Ledger
+
+	eval *internal.BlockEvaluator
+}
+
+func (dl DoubleLedger) Close() {
+	dl.generator.Close()
+	dl.validator.Close()
+}
+
+// NewDoubleLedger creates a new DoubleLedger with the supplied balances and consensus version.
+func NewDoubleLedger(t *testing.T, balances bookkeeping.GenesisBalances, cv protocol.ConsensusVersion) DoubleLedger {
+	g := newTestLedgerWithConsensusVersion(t, balances, cv)
+	v := newTestLedgerFull(t, balances, cv, g.GenesisHash())
+	return DoubleLedger{t, g, v, nil}
+}
+
+func (dl *DoubleLedger) beginBlock() *internal.BlockEvaluator {
+	dl.eval = nextBlock(dl.t, dl.generator)
+	return dl.eval
+}
+
+func (dl *DoubleLedger) txn(tx *txntest.Txn, problem ...string) {
+	dl.t.Helper()
```

Comment:
@jannotti Golang newbie question:  What operations / scenarios does `t.Helper()` impact?

I read https://pkg.go.dev/testing@go1.16.9#T.Helper and I'm still puzzled.  I incorrectly thought the stacktrace on failure is mutated.  

If it's an involved discussion, we can do it live.  It's a low urgency concern.

Reply:
By marking a test t.Helper(), failures are reported at the parent, rather than in that function.

Reply:
Got it - thanks for the explanation. Resolving.